### DISCUSSION TITLE
How do you count the substructures in a molecule?

### DISCUSSION THREAD

**Discussion by okikosuke** [rdkit/rdkit#5366] [Q&A] [answered]
I want to count the number of -CH2 and -CH3 in a molecule.

Suppose we have the following compounds.
SMILES=CC/C=C/C1=CC=CC=C1
If we ignore the benzene ring, we want to count the number of -CH2 in this compound as 3, the number of -CH3 as 1, and so on. You can use RDKit to determine if there is a substructure or not.I would also like to know how to handle benzene rings such that I can count the number of -CH2- and =CH- in each benzene ring as three.

**Comment by greglandrum** [comment]
Hi @okikosuke, the molecule has a method `GetSubstructMatches()` which returns a list of all the matches of the substructure. I think that does what you are looking for.

An aside; because benzene rings are converted to the aromatic form, you will not find any matches for `=CH-` in a benzene ring.

**Comment by Pakman450** [comment]
There is one way that I had some success with, but it requires good knowledge on how to create SMARTS patterns. 
[smarts.plus](https://smarts.plus/)  is a website where you can visualize what your SMARTS patterns is going to require to have a match. [(Research paper if you want it)](https://pubmed.ncbi.nlm.nih.gov/32997890/) The more lenient your definitions are, the more open the matches you are going to have with your patterns. SMARTS strings can be very particular on how YOU define what is considered as a match. This is why I would also read this [documentation](https://www.daylight.com/dayhtml/doc/theory/theory.smarts.html) so you can learn how to form your own SMARTS strings to your liking. Then visualize your SMARTS patterns in smarts.plus to see if you defined well enough for your purposes. 

an example of the visualization of SMARTS patterns are shown here (top left is the SMARTS pattern in question):

![Screen Shot 2022-06-14 at 1 20 10 PM](https://user-images.githubusercontent.com/69532598/173638351-37df4d81-573c-419d-9e91-c48ebf15dfce.png)

Then, as @greglandrum said, you can use `GetSubstructMatches()` to finally get those matches!

Good luck!

**Comment by okikosuke** [comment]
Hi @greglandrum @Pakman450 
Thanks for the reply. Thanks to you, I understand much better. However, I am having a problem that I can't solve. Please let me know if you know how to solve it.

The SMILES of 1-substituted benzene is found to be "c1ccccc1*". 

```:python
x = Chem.MolFromSmarts("c1ccccc1*")
x
```

![image](https://user-images.githubusercontent.com/74843794/174039586-df361ead-109c-4cd3-b6c7-25b587125c90.png)

Consider the following example.

```:python
molecule = Chem.MolFromSmiles("CC(O)CC/C=C/C(C1=CC=CC=C1)=O")
molecule
```

![image](https://user-images.githubusercontent.com/74843794/174037858-26083771-6359-437c-9463-7abe11e5e296.png)

```:python
molecule.GetSubstructMatches(x)
```

```:output
((9, 10, 11, 12, 13, 8, 7),)
```

![image](https://user-images.githubusercontent.com/74843794/174038067-add26ae3-7a84-4506-91cb-e05ce894cb68.png)

Thus, we were indeed able to find the substructure of a monosubstituted benzene from the molecule. However, I want to distinguish between mono-substituted benzene and di-substituted benzene.Consider the following 2-substituted benzene.

```:python
molecule2 = Chem.MolFromSmiles("O=C(OCC(C)(C)C)C1=CC=CC(C(O)=O)=C1")
molecule2
```

![image](https://user-images.githubusercontent.com/74843794/174038508-ea9a80d2-8fb7-40d9-a3f2-3dd2da45b4df.png)

```:python
molecule2.GetSubstructMatches(x)
```

```:output
((9, 10, 11, 12, 16, 8, 1), (11, 10, 9, 8, 16, 12, 13))
```

![image](https://user-images.githubusercontent.com/74843794/174038791-607a4926-7bab-4c3d-b698-b960b099db10.png)

Thus, the substructure of a 1-substituted benzene would also be searched for in a 2-substituted benzene. I would like the substructure of 1-substituted benzene to be detected only in 1-substituted benzene.
Similarly, the substructure of mono-substituted benzene should be detected in naphthalene.

What should I do? Please let me know the solution.

**Comment by greglandrum** [comment, marked as answer]
@okikosuke it souds like you want more restricted substructure searches. For this you need to either make your SMARTS query more specific or use AdjustQueryProperties to tune the query.
Here's a simple example:
```
In [8]: qry = Chem.MolFromSmarts('*c1ccccc1')

In [9]: qry2 = Chem.AdjustQueryProperties(qry)

In [10]: m1 = Chem.MolFromSmiles('Cc1ccccc1')

In [11]: m2 = Chem.MolFromSmiles('Cc1c(C)cccc1')

In [12]: m1.HasSubstructMatch(qry)
Out[12]: True

In [13]: m1.HasSubstructMatch(qry2)
Out[13]: True

In [14]: m2.HasSubstructMatch(qry)
Out[14]: True

In [15]: m2.HasSubstructMatch(qry2)
Out[15]: False
```

here's a SMARTS showing what AdjustQueryProperties did:
```
In [16]: Chem.MolToSmarts(qry2)
Out[16]: '*[c&D3]1[c&D2][c&D2][c&D2][c&D2][c&D2]1'
```

There's more information on what you can do with AdjustQueryProperties() here: http://rdkit.blogspot.com/2016/07/tuning-substructure-queries-ii.html

**Comment by okikosuke** [comment]
@greglandrum 
Thank you for your reply. I'll try again with your idea.
